doc/contribute/committer_expectations.rst

```diff
@@ -0,0 +1,165 @@
+.. _committer-expectations:
+
+Committer Expectations
+######################
+
+Overview
+********
+
+The Zephyr project encourages you to submit changes as small pull-requests.
+Small pull requests (PRs) have the following benefits:
+
+- Reviewed more quickly and reviewed more thoroughly. It's easier for reviewers
+  to set aside a few minutes to review small changes several times than it is to
+  allocate a 30 minute block to review a large PR.
+
+- Less wasted work if the PR if reviewers or maintainers reject the direction of
+  the change.
+
+- Easier to rebase and merge. Smaller PRs are less likely to conflict with other
+  changes in the tree.
+
+- Easier to rollback if the PR breaks functionality.
+
+
+Defining Small PRs
+==================
+
+- Small PRs should encompass one self-contained logical change.
+
+- When adding a new feature or API, the PR should address only one part of the
+  feature. In this case create an :ref:`RFC proposal <rfcs>` [#rfc]_ to describe
+  the additional parts of the feature for reviewers.
+
+- PRs should include tests for any added or changed lines of code.
```

Comment:
```suggestion
- PRs should include tests or samples for any added or changed lines of code.
```

Reply:
Is there a particular goal you have in mind by trying to require testing here? Many contributions to Zephyr come from one off small driver additions/features/bugfixes from contributors working on a product. There's already the remarkably high barrier to entry where many PRs die today of simply passing git/code linting steps. Adding yet another brick on the wall feels like we'll lose those first time/small contributors even faster. Sensors are especially prone to this sort of thing. There's over a 100 sensor drivers that more or less are community contributed and tested, with the project providing samples to build against and run.

Speaking of which if a PR contains a bug fix for a piece of hardware what is the real expectation here? If we look at i2c drivers for example very few are actually being tested on hardware given the i2c test case requirement of having a specific external IC connected to it.

Testing by building the driver in a sample or test is about the same, its a smoke test to make sure at least the code builds which is 90% of what we get with CI today when you exclude simulators.

If you consider that, samples and tests are about equivalent in that regard.

Reply:
IMO, the Zephyr project should move towards 100% test coverage, including drivers.  I think the expectations for tests should be:
1. When fixing a hardware-agnostic bug, PR authors should create a new test that passes with the fix applied, but fails without the fix.
2. When fixing a hardware specific bug (SoC driver, peripheral driver, etc), the PR author should fix the corresponding tests/emulators as well.  If the driver doesn't currently have any tests or emulation, this requirement is waived.  I agree it isn't reasonable to hold up a bug fix and require the PR author to write a new emulator when one doesn't exist.
3. When introducing a new driver, one should also contribute an emulator and tests for the driver.  The PR author has already done a deep dive into the datasheet and has the expertise to write an emulator at the same time.  I fully expect that there will be latent bugs introduced because the PR author writes both the driver and emulator.  However, once the driver is run on real hardware any driver bug fixes will require a fix to the emulator (see previous comment).